packages/flutter/lib/src/cupertino/text_field.dart

```diff
@@ -0,0 +1,657 @@
+
+  @override
+  void deactivate() {
+    super.deactivate();
+  }
+
+  bool _shouldShowAttachment({
+    OverlayVisibilityMode attachment,
+    bool hasText,
+  }) {
+    switch (attachment) {
+      case OverlayVisibilityMode.never:
+        return false;
+      case OverlayVisibilityMode.always:
+        return true;
+      case OverlayVisibilityMode.editing:
+        return hasText;
+      case OverlayVisibilityMode.notEditing:
+        return !hasText;
+    }
+    assert(false);
+    return null;
+  }
+
+  Widget _addTextDependentAttachments(Widget editableText) {
+    // If there are no surrounding widgets, just return the core editable text
+    // part.
+    if (widget.placeholder == null &&
+        widget.clearButtonMode == OverlayVisibilityMode.never &&
+        widget.prefix == null &&
+        widget.suffix == null) {
+      return editableText;
+    }
+
+    // Otherwise, listen to the current state of the text entry.
+    return ValueListenableBuilder<TextEditingValue>(
+      valueListenable: _effectiveController,
+      child: editableText,
+      builder: (BuildContext context, TextEditingValue text, Widget child) {
+        final List<Widget> rowChildren = <Widget>[];
+
+        // Insert a prefix at the front if the prefix visibility mode matches
+        // the current text state.
+        if (widget.prefix != null &&
+            _shouldShowAttachment(
+              attachment: widget.prefixMode,
+              hasText: text.text.isNotEmpty,
+            )
+        ) {
+          rowChildren.add(widget.prefix);
+        }
+
+        final List<Widget> stackChildren = <Widget>[];
+
+        // In the middle part, stack the placeholder on top of the main EditableText
+        // if needed.
+        if (widget.placeholder != null && text.text.isEmpty) {
+          stackChildren.add(
+            Padding(
+              padding: widget.padding,
+              child: Text(
+                widget.placeholder,
+                maxLines: 1,
+                overflow: TextOverflow.ellipsis,
+                style: widget.style.merge(
+                  const TextStyle(
+                    color: _kInactiveTextColor,
+                    fontWeight: FontWeight.w300,
+                  ),
+                ),
+              ),
+            ),
+          );
+        }
+
+        rowChildren.add(Expanded(child: Stack(children: stackChildren..add(child))));
+
+        // First add the explicit suffix if the suffix visibility mode matches.
+        if (widget.suffix != null &&
+            _shouldShowAttachment(
+              attachment: widget.suffixMode,
+              hasText: text.text.isNotEmpty,
+            )
+        ) {
+          rowChildren.add(widget.suffix);
+        // Otherwise, try to show a clear button if its visibility mode matches.
+        } else if (
+          _shouldShowAttachment(
+            attachment: widget.clearButtonMode,
+            hasText: text.text.isNotEmpty,
+          )
+        ) {
```

Comment:
i guess technically this is flutter-style wrapping but it's very confusing to look at.

Reply:
same